File: client/src/main/java/nl/altindag/client/service/DispatchRebootService.java
package nl.altindag.client.service;

import dispatch.Http;
import nl.altindag.client.ClientType;
import nl.altindag.client.model.ClientResponse;
import org.asynchttpclient.RequestBuilder;
import org.asynchttpclient.Response;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.concurrent.TimeUnit;

import static nl.altindag.client.ClientType.DISPATCH_REBOOT_HTTP_CLIENT;
import static nl.altindag.client.Constants.HEADER_KEY_CLIENT_TYPE;

@Service
public class DispatchRebootService implements RequestService {

    private static final int TIMEOUT_AMOUNT_IN_SECONDS = 5;

    private final Http httpClient;

    @Autowired
    public DispatchRebootService(Http httpClient) {
        this.httpClient = httpClient;
    }

    @Override
    public ClientResponse executeRequest(String url) throws Exception {
        RequestBuilder requestBuilder = new RequestBuilder()
                .setUrl(url)
                .setHeader(HEADER_KEY_CLIENT_TYPE, getClientType().getValue());

        Response response = httpClient.client().executeRequest(requestBuilder)
                .toCompletableFuture()
                .get(TIMEOUT_AMOUNT_IN_SECONDS, TimeUnit.SECONDS);

        return new ClientResponse(response.getResponseBody(), response.getStatusCode());
    }

    @Override
    public ClientType getClientType() {
        return DISPATCH_REBOOT_HTTP_CLIENT;
    }

}

File: client/src/main/java/nl/altindag/client/service/DispatchRebootService.scala
/*
 * Copyright 2018 Thunderberry.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package nl.altindag.client.service

import dispatch.Defaults._
import dispatch._
import nl.altindag.client.ClientType
import nl.altindag.client.ClientType.DISPATCH_REBOOT_HTTP_CLIENT
import nl.altindag.client.Constants.HEADER_KEY_CLIENT_TYPE
import nl.altindag.client.model.ClientResponse
import nl.altindag.ssl.SSLFactory
import nl.altindag.ssl.netty.util.NettySslUtils
import org.springframework.context.annotation.Bean
import org.springframework.stereotype.{Component, Service}

import scala.language.postfixOps

@Service
class DispatchRebootService(httpClient: Http) extends RequestService {

  override def executeRequest(uri: String): ClientResponse = {
    val request: Req = url(uri).GET addHeader (HEADER_KEY_CLIENT_TYPE, getClientType.getValue)

    httpClient(request)
      .map(response => new ClientResponse(response.getResponseBody, response.getStatusCode))
      .apply()
  }

  override def getClientType: ClientType = DISPATCH_REBOOT_HTTP_CLIENT

}

@Component
class DispatchRebootHttpClientConfig {

  @Bean
  def createDispatchRebootHttpClient(sslFactory: SSLFactory): Http = {
    val sslContext = NettySslUtils.forClient(sslFactory).build
    Http.withConfiguration(builder => builder.setSslContext(sslContext))
  }

}
